### ISSUE TITLE
Subscription filter does not have a bound instance when dependency tree is not static

### ISSUE THREAD

**Issue by greenreign** [nestjs/graphql#1598]
## I'm submitting a...

<pre><code>
[ ] Regression 
[X] Bug report
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead post your question on Stack Overflow.
</code></pre>

## Current behavior
When your resolver has any dependency, pipe, guard etc which is `Scope.REQUEST` scoped then subscription filters will not have a valid bounded instance.  The instance will be of the correct type but will be empty and not contain any of your dependencies.  This also correlates with the `isTreeStatic` property being `false`

## Expected behavior

Following [the pattern just below here](https://github.com/nestjs/docs.nestjs.com/blob/master/content/graphql/subscriptions.md#mutating-subscription-payloads) to access injected dependencies inside my `filter` function.
```
If you need to access injected providers (e.g., use an external service to validate the data), use the following construction.

@Subscription(returns => Comment, {
  resolve(this: AuthorResolver, value) {
    // "this" refers to an instance of "AuthorResolver"
    return value;
  }
})
commentAdded() {
  return pubSub.asyncIterator('commentAdded');
}
The same construction works with filters:

@Subscription(returns => Comment, {
  filter(this: AuthorResolver, payload, variables) {
    // "this" refers to an instance of "AuthorResolver"
    return payload.commentAdded.title === variables.title;
  }
})
commentAdded() {
  return pubSub.asyncIterator('commentAdded');
}
```
Should result in an instance where I will have access to my injected dependencies of the resolver

## Minimal reproduction of the problem with instructions
https://github.com/coast-app/test-gql-subscriptions

## What is the motivation / use case for changing the behavior?
I think it's a bug and should perform as documented.

## Environment

<pre><code>
Nest version: X.Y.Z
@nestjs/graphql@7.11.0
 
For Tooling issues:
- Node version: 14.15.3
- Platform:  Windows

Others:

</code></pre>

**Comment by jmcdo29** [comment]
I believe this is a similar reason as why websocket gateways cannot be request scoped

**Comment by greenreign** [comment]
Feels like it should be fixable.  If anyone has some guidance I'd be happy to work on it.  I looked at the code and I'm not sure what would be the best way to attack it though.  We have access to `moduleRef` in the `resolvers-explorer` and `base-explorer` but it's unclear to me what exactly is going on.  Does having a REQUEST scoped pipe in a resolver mean that my resolver is essentially request scoped?

**Comment by greenreign** [comment]
@jmcdo29 See Q above. Should I just move my `@Subscription` out of the resolver into it's own `AppSubscriptionsResolver` which does not rely on anything request-scoped?

**Comment by jmcdo29** [comment]
I'd probably move it, yeah. I don't _think_ a request scoped enhancer should affect the resolver class, but I'll wait and see what Kamil says about that

**Comment by kamilmysliwiec** [comment]
> I believe this is a similar reason as why websocket gateways cannot be request scoped

That's correct. Subscription handlers shouldn't be request-scoped as it doesn't make any sense to do so.

Still, instead of passing the `undefined` value as an instance reference to filters/transformers, we should rather instruct the developer (by logging down the error?) that doing this (request-scoping subscription handlers) isn't the right way, to improve the overall developer experience.

**Comment by kamilmysliwiec** [comment]
Would you like to create a PR for this issue?

**Comment by greenreign** [comment]
@kamilmysliwiec It's still not clear.  What makes this subscription handler request-scoped? I think it's the Pipe on a different method/resolver.  Is that expected?

**Comment by kamilmysliwiec** [comment]
Nest doesn't have a concept of request-scoped methods. If anything within a class is registered as a request-scoped enhancer/provider, then the class itself becomes request-scoped as well.
